HeterogeneousCore/CUDAServices/src/CUDAService.cc

```diff
@@ -60,6 +60,50 @@ CUDAService::CUDAService(edm::ParameterSet const& iConfig, edm::ActivityRegistry
 
   edm::LogInfo("CUDAService") << "CUDAService fully initialized";
   enabled_ = true;
+
+  // see the documentation of cudaDeviceSetLimit for more information
+  auto limits = iConfig.getUntrackedParameter<edm::ParameterSet>("limits");
+  size_t request, value;
+
+  // cudaLimitPrintfFifoSize controls the size in bytes of the shared FIFO used by the
+  // printf() device system call.
+  request = limits.getUntrackedParameter<unsigned int>("printfFifoSize");
+  if (request > 0) {
+    cudaCheck(cudaDeviceSetLimit(cudaLimitPrintfFifoSize, request));
+    cudaCheck(cudaDeviceGetLimit(&value, cudaLimitPrintfFifoSize));
+    edm::LogInfo("CUDAService") << "cudaLimitPrintfFifoSize requested " << request << ", set to " << value;
```

Comment:
Overall looks good to me. It is not clear to me, however, if these settings apply automatically to all devices, or only the current device. The manual (*) doesn't explicitly say it would be a global setting, but on the other hand, it says for many other functions "for the current device", so maybe it is a global setting.

(*) https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html

Reply:
Good point, I'll see if I can somehow check empirically.

Reply:
At least stack seems to be device specific. The following
```cpp
#include <iostream>
#include <cuda_runtime.h>
#include <cuda.h>

int main(void) {
  int deviceCount = 0;
  auto ret = cudaGetDeviceCount( &deviceCount );

  std::cout << "Devices " << deviceCount << std::endl;

  size_t value;
  cudaDeviceGetLimit(&value, cudaLimitStackSize);
  std::cout << "Stack size " << value << std::endl;

  value = 2048;
  cudaDeviceSetLimit(cudaLimitStackSize, value);
  cudaDeviceGetLimit(&value, cudaLimitStackSize);
  std::cout << "Set stack size " << value << std::endl;

  cudaSetDevice(1);
  cudaDeviceGetLimit(&value, cudaLimitStackSize);
  std::cout << "Device 1, stack size " << value << std::endl;

  return 0;
}
```
ran on the multi-GPU CMG machine gives
```
$ CUDA_VISIBLE_DEVICES=1,2 ./foo
Devices 2
Stack size 1024
Set stack size 2048
Device 1, stack size 1024
```